fetch/src/main/scala/fetch.scala

```diff
@@ -156,7 +156,10 @@ object `package` {
       y: BlockedRequest[F]
   ): BlockedRequest[F] =
     (x.request, y.request) match {
-      case (a @ FetchOne(aId, ds), b @ FetchOne(anotherId, _)) =>
+      case (a: FetchOne[Any, Any], b: FetchOne[Any, Any]) =>
```

Comment:
This is a bit unfortunate. I can't claim to fully understand all this code, but as far as I can tell, it was taking advantage of bugs in the Scala compiler's GADT support in order to compile?

I had a quick look but I have a feeling that any attempt to address this properly would become a rewrite of most of the library. We could at least try to suppress the erasure warnings?

Reply:
Something we can try if the `@` operator doesn't work is to rewrap the pattern again if needed?

```scala
      case (FetchOne(aId, ds), FetchOne(anotherId, ds2)) =>
        val a = FetchOne(aId, ds)
        val b = FetchOne(anotherId, ds2)
```

Reply:
I tried that but I am getting the same compilation error:
```
[error] -- [E007] Type Mismatch Error: /Users/damien/code/fetch/fetch/src/main/scala/fetch.scala:167:48
[error] 167 |          val combined   = combineIdentities(a, b)
[error]     |                                                ^
[error]     |                      Found:    (b : fetch.package.FetchOne[I$2, A$2])
[error]     |                      Required: fetch.FetchQuery[I$1, A$1]
```
The compiler wants to prove that the type parameters of `FetchOne` a and b are the same but it can't. We lose those parameters when we use FetchRequest.

Reply:
I'll try to see if we can keep track of the `I` and `A` type params along the way. I gave it a quick go now and it doesn't seem so straightforward.

Reply:
I spent some time this afternoon trying to keep track of the `I` and avoid the `FetchOne[Any, Any]`. However, I end up bubbling that type parameter all the way up in the API (in the pure and apply method of the `Fetch` object) which I don't think we want to do.

I am not sure there is a way around using `Any` here. If you look at how the `Fetch` are created, we widen to `Any` in  the fetch constructor methods (`apply`, `optional`) and then cast with `asInstanceOf[A]`. Looks like this is where we loose the `I` we would need here.

Also the monad instance for fetch, `implicit def fetchM[F[_]: Monad]: Monad[Fetch[F, *]]` would have to be reworked if we push the `I` all the way up which seems complicated.